master/internal/stream/system.go

```diff
@@ -191,6 +183,100 @@ func writeAll(socketLike WebsocketLike, msgs []interface{}) error {
 	return nil
 }
 
+// processStream processes as startup message, then streams live updates until either:
+// - another startup message arrives, in which case it returns it or
+// - the streamer is closed gracefully, in which case it returns nil, or
+// - an error occurs.
+func (ps *PublisherSet) processStream(
+	ctx context.Context,
+	streamer *stream.Streamer,
+	user model.User,
+	startupMsg StartupMsg,
+	startups *[]StartupMsg,
+	socket WebsocketLike,
+) (
+	nextStartup *StartupMsg, err error,
+) {
+	// create new subscription set
+	ss, err := NewSubscriptionSet(ctx, streamer, ps, user, startupMsg.Subscribe)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating subscription set")
+	}
+	defer ss.UnregisterAll()
+
+	// startup subscription set
+	msgs := []interface{}{}
+	offlineMsgs, err := ss.Startup(ctx, user, startupMsg)
+	if err != nil {
+		return nil, errors.Wrapf(err, "gathering startup messages")
+	}
+	msgs = append(msgs, offlineMsgs...)
+
+	// always include a sync message
+	syncMsg := SyncMsg{SyncID: startupMsg.SyncID}.toPreparedMessage()
+	msgs = append(msgs, syncMsg)
+
+	// write offline msgs to the websocket
+	err = writeAll(socket, msgs)
+	if err != nil {
+		// don't log broken pipe errors
+		if !websocket.IsCloseError(err, websocket.CloseAbnormalClosure) {
+			log.Debugf("unable to handle startup message: %s", err)
+		}
+		return nil, errors.Wrapf(err, "error handling startup message")
+	}
+
+	// startup done, begin streaming supported online events:
+	//	- insertions
+	//	- updates
+	//	- deletions
+	//	- fallin
+	//	- fallout
+	// note: online appearances and disappearances are not supported; we'll detect those
+	// situtations and break the connection to the streaming clients
+
+	// waitForSomething Returns a tuple of (first-startup-msg, msgs, closed)
+	waitForSomething := func() (*StartupMsg, []interface{}, bool) {
+		streamer.Cond.L.Lock()
+		defer streamer.Cond.L.Unlock()
+
+		if len(*startups) == 0 && len(streamer.Msgs) == 0 && streamer.Closed {
```

Comment:
Why did you switch from the `for` loop I suggested to an `if` statement here?

Reply:
Good catch, it got swapped because I missed it; using an if statement doesn't make any since for this case (waiting in a while is general the way to go)